File: Server~/src/tools/assetTools/moveAssetTool.ts
import * as z from 'zod';
import { Logger } from '../../utils/logger.js';
import { McpUnity } from '../../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'move_asset';
const toolDescription = 'Moves assets (files or folders) from source locations to a destination in the project';
const paramsSchema = z.object({
  sourcePaths: z.array(z.string()).describe('The source path(s) of the asset(s) to move'),
  destPath: z.string().describe('The destination path where the asset will be moved to')
});

export function registerMoveAssetTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);
      
  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const { sourcePaths, destPath } = params;
        
  if (!sourcePaths || !destPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      'Required parameters "sourcePaths" and "destPath" must be provided'
    );
  }

  if (sourcePaths.length === 0) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      'At least one source path must be provided'
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: {
      sourcePaths,
      destPath
    }
  });
  
  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to move assets to ${destPath}`
    );
  }

  return {
    content: [{
      type: response.type || 'text',
      text: response.message,
      errors: response.errors
    }]
  };
}

File: Editor/Tools/AssetTools/MoveAssetTool.cs
using UnityEngine;
using UnityEditor;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using McpUnity.Tools;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    public class MoveAssetTool : McpToolBase
    {
        public MoveAssetTool()
        {
            Name = "move_asset";
            Description = "Moves assets (files or folders) from source locations to a destination in the project";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            JToken sourcePathsToken = parameters["sourcePaths"];
            string destPath = parameters["destPath"]?.ToObject<string>();

            if (sourcePathsToken == null || string.IsNullOrEmpty(destPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameters 'sourcePaths' and 'destPath' must be provided",
                    "validation_error"
                );
            }

            List<string> sourcePaths;
            if (sourcePathsToken.Type == JTokenType.Array)
            {
                sourcePaths = sourcePathsToken.ToObject<List<string>>();
            }
            else
            {
                sourcePaths = new List<string> { sourcePathsToken.ToObject<string>() };
            }

            if (!sourcePaths.Any())
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "At least one source path must be provided",
                    "validation_error"
                );
            }

            destPath = AssetUtils.EnsureAssetPath(destPath);

            int results = 0;
            Dictionary<string, string> errors = new Dictionary<string, string>();

            foreach (string sourcePath in sourcePaths)
            {
                string normalizedSourcePath = AssetUtils.EnsureAssetPath(sourcePath);
                
                if (!AssetUtils.CheckAssetExists(normalizedSourcePath))
                {
                    errors.Add(normalizedSourcePath, "Source asset not found");
                    continue;
                }

                string finalDestPath = Path.Combine(destPath, Path.GetFileName(normalizedSourcePath));

                if (AssetUtils.CheckAssetExists(finalDestPath))
                {
                    errors.Add(finalDestPath, "Asset already exists at destination");
                    continue;
                }

                try 
                {
                    AssetUtils.EnsureDirectoryExists(Path.GetDirectoryName(finalDestPath));

                    string validationError = AssetDatabase.ValidateMoveAsset(normalizedSourcePath, finalDestPath);
                    if (!string.IsNullOrEmpty(validationError))
                    {
                        errors.Add(normalizedSourcePath, validationError);
                        continue;
                    }

                    string error = AssetDatabase.MoveAsset(normalizedSourcePath, finalDestPath);
                    
                    if (string.IsNullOrEmpty(error))
                    {
                        ++results;
                    }
                    else
                    {
                        errors.Add(normalizedSourcePath, error);
                    }
                }
                catch (System.Exception ex)
                {
                    errors.Add(normalizedSourcePath, ex.Message);
                }
            }

            return new JObject
            {
                ["type"] = "text",
                ["message"] = $"Moved assets: {results} successfully, {errors.Count} failed",
                ["errors"] = errors.Count > 0 ? JObject.FromObject(errors) : null
            };
        }
    }
}
